source/common/config/datasource.h

```diff
@@ -78,12 +79,11 @@ class RemoteAsyncDataProvider : public Config::DataFetcher::RemoteDataFetcherCal
 
   // Config::DataFetcher::RemoteDataFetcherCallback
   void onFailure(Config::DataFetcher::FailureReason failure) override {
+    ENVOY_LOG(debug, "Failed to fetch remote data. Failure reason: {}", enumToInt(failure));
     if (allow_empty_) {
       callback_(EMPTY_STRING);
+      // Only raise ready when caller allows empty.
```

Comment:
Won't this potentially get stuck forever during startup with the missing exception below? Don't we need to restart the fetch, have a timeout, etc.?

Reply:
The remote data source already has a timeout. I think we need to add the retry policy in the following PR.

Reply:
It will get stuck forever here. If `init_target` raises ready anyway, it could be dangerous. E.g., for WASM, it's possible that the wasm vm is null during the the request path.

The problem for throwing exception is that we cannot know where will the exception be caught.

Reply:
But at least then it will crash and let us know that something is obviously busted vs. getting stuck. Can we figure out the right retry/timeout, etc. in this PR? Thank you.

/wait

Reply:
I am going to do following two things:

1) add retry policy for remote data source. HTTP router has retry policy https://github.com/envoyproxy/envoy/blob/6eeced62c2df393585e83fc0377f81e332ffc2c2/api/envoy/api/v2/route/route.proto#L878

but we don't need some fields, e.g., `RetryHostPredicate`. I would prefer to add a simplified retry policy. WDYT?

2) raise `ready` event after retry anyway (it could crash the Envoy)

Reply:
Yeah I think it's fine to have a simplified retry policy, but please make sure to use a jittered backoff strategy which I think already has common protos in flight. cc @htuch 

I think my main concern is that we don't block envoy startup. If you want to land this PR without retry, I think in the failure case you need to still continue startup.

Reply:
SGTM. I am going to raise `ready` anyway in this PR to continue startup. Opened an issue for retry policy.